File: 3483.unique-3-digit-even-numbers/solution.ts
/*
 * @lc app=leetcode id=1910 lang=typescript
 *
 * 3483. Unique 3-Digit Even Numbers
 * https://leetcode.com/problems/unique-3-digit-even-numbers
 */

// @lc code=start
function totalNumbers(digits: number[]): number {
    const freq = new Array(10).fill(0);
    for (const d of digits) {
        freq[d]++;
    }
    let count = 0;
    for (let units = 0; units < 10; units += 2) {
        if (freq[units] === 0) continue;
        freq[units]--;
        for (let tens = 0; tens < 10; tens++) {
            if (freq[tens] === 0) continue;
            freq[tens]--;
            count += freq.slice(1, 10).filter(e => e > 0).length;
            freq[tens]++;
        }
        freq[units]++;
    }
    return count;
};
// @lc code=end

console.log(totalNumbers([1,2,3,4]));
console.log(totalNumbers([0,2,2]));

File: 3483.unique-3-digit-even-numbers/solution.py
#
# @lc app=leetcode id=1910 lang=python3
#
# 3483. Unique 3-Digit Even Numbers
# https://leetcode.com/problems/unique-3-digit-even-numbers
#


# @lc code=start
from typing import List


class Solution:
    def totalNumbers(self, digits: List[int]) -> int:
        freq = [0] * 10
        for d in digits:
            freq[d] += 1

        count = 0
        for units in range(0, 10, 2):
            if not freq[units]:
                continue
            freq[units] -= 1
            for tens in range(10):
                if not freq[tens]:
                    continue
                freq[tens] -= 1
                count += sum(1 for h in range(1, 10) if freq[h])
                freq[tens] += 1
            freq[units] += 1
        return count
            

if __name__ == "__main__":
    print(Solution().totalNumbers([1,2,3,4]))
    print(Solution().totalNumbers([0,2,2]))

# @lc code=end